packages/webpack/src/builder.js

```diff
@@ -0,0 +1,284 @@
+import upath from 'upath'
+import consola from 'consola'
+
+import {
+  r,
+  parallel,
+  sequence
+} from '@nuxt/common'
+
+import { ClientConfig, ServerConfig, PerfLoader } from './config'
+
+export default class WebpackBuilder {
+  constructor(context) {
+    this.context = context
+    this.nuxt = context.nuxt
+    this.options = context.nuxt.options
+    this.isStatic = context.isStatic
+    this.plugins = context.plugins
+    // Fields that set on build
+    this.compilers = []
+    this.compilersWatching = []
+    this.webpackDevMiddleware = null
+    this.webpackHotMiddleware = null
+    this.watchers = {
+      files: null,
+      custom: null,
+      restart: null
+    }
+    this.perfLoader = null
+
+    // Initialize shared FS and Cache
+    if (this.options.dev) {
+      this.mfs = new MFS()
+    }
+  }
+
+  async build() {
+    this.perfLoader = new PerfLoader(this.options)
+
+    const compilersOptions = []
+
+    // Client
+    const clientConfig = new ClientConfig(this).config()
+    compilersOptions.push(clientConfig)
+
+    // Server
+    let serverConfig = null
+    if (this.options.build.ssr) {
+      serverConfig = new ServerConfig(this).config()
+      compilersOptions.push(serverConfig)
+    }
+
+    for (const p of this.plugins) {
+      // Client config
+      if (!clientConfig.resolve.alias[p.name]) {
+        clientConfig.resolve.alias[p.name] = p.src
+      }
+
+      // Server config
+      if (serverConfig && !serverConfig.resolve.alias[p.name]) {
+        // Alias to noop for ssr:false plugins
+        serverConfig.resolve.alias[p.name] = p.ssr ? p.src : './empty.js'
+      }
+    }
+
+    // Configure compilers
+    this.compilers = compilersOptions.map((compilersOption) => {
+      const compiler = webpack(compilersOption)
+
+      // In dev, write files in memory FS
+      if (this.options.dev) {
+        compiler.outputFileSystem = this.mfs
+      }
+
+      return compiler
+    })
+
+    // Warmup perfLoader before build
+    if (this.options.build.parallel) {
+      consola.info('Warming up worker pools')
+      this.perfLoader.warmupAll()
+      consola.success('Worker pools ready')
+    }
+
+    // Start Builds
+    const runner = this.options.dev ? parallel : sequence
+
+    await runner(this.compilers, (compiler) => {
+      return this.webpackCompile(compiler)
+    })
+  }
+
+  webpackCompile(compiler) {
+    return new Promise(async (resolve, reject) => {
+      const name = compiler.options.name
+
+      await this.nuxt.callHook('build:compile', { name, compiler })
+
+      // Load renderer resources after build
+      compiler.hooks.done.tap('load-resources', async (stats) => {
+        await this.nuxt.callHook('build:compiled', {
+          name,
+          compiler,
+          stats
+        })
+
+        // Reload renderer if available
+        this.nuxt.renderer.loadResources(this.mfs || fs)
+
+        // Resolve on next tick
+        process.nextTick(resolve)
```

Comment:
As we probably lose history, would you please add a hint comment to the original issue?